pkg/gatewayserver/gatewayserver.go

```diff
@@ -399,25 +430,72 @@ func (gs *GatewayServer) handleUpstream(conn *io.Connection) {
 		}
 	}
 
-	defer wg.Wait()
+	hosts := make([]*upstreamHost, 0, len(gs.forward))
+	for name, prefixes := range gs.forward {
+		passDevAddr := func(devAddr types.DevAddr) bool {
+			for _, prefix := range prefixes {
+				if devAddr.HasPrefix(prefix) {
+					return true
+				}
+			}
+			return false
+		}
+		if name == "" {
+			// Cluster Network Server; filter based on DevAddr and pass all join-requests.
+			hosts = append(hosts, &upstreamHost{
+				name: "cluster",
+				handler: func(ids *ttnpb.EndDeviceIdentifiers) upstreamHandler {
+					if ids != nil && ids.DevAddr != nil && !passDevAddr(*ids.DevAddr) {
+						return nil
+					}
+					ns := gs.GetPeer(ctx, ttnpb.PeerInfo_NETWORK_SERVER, ids)
+					if ns == nil {
+						logger.Warn("Cluster Network Server unavailable for upstream traffic")
+						return nil
+					}
+					return ttnpb.NewGsNsClient(ns.Conn())
+				},
+				callOpts: []grpc.CallOption{gs.WithClusterAuth()},
+				handleCh: make(chan upstreamItem),
+			})
+		} else {
+			// Packet Broker; filter based on DevAddr and filter all join-requests.
+			// TODO: Offload traffic to Packet Broker (https://github.com/TheThingsNetwork/lorawan-stack/issues/671)
+		}
+	}
+
+	for _, host := range hosts {
+		defer host.handleWg.Wait()
+	}
+
 	for {
-		var item interface{}
+		var val interface{}
 		select {
 		case <-ctx.Done():
 			return
-		case item = <-conn.Up():
-		case item = <-conn.Status():
-		case item = <-conn.TxAck():
+		case val = <-conn.Up():
+		case val = <-conn.Status():
+		case val = <-conn.TxAck():
 		}
-		select {
-		case handleCh <- item:
-		default:
-			if atomic.LoadInt32(&handlers) < maxUpstreamHandlers {
-				atomic.AddInt32(&handlers, 1)
-				wg.Add(1)
-				go handleFn()
+		for _, host := range hosts {
+			item := upstreamItem{
+				val:  val,
+				host: host,
+			}
+			select {
+			case host.handleCh <- item:
+			default:
+				if atomic.LoadInt32(&host.handlers) < maxUpstreamHandlers {
+					atomic.AddInt32(&host.handlers, 1)
+					host.handleWg.Add(1)
+					go handleFn(host)
+				}
+				select {
+				case host.handleCh <- item:
+				case <-time.After(upstreamHandlerBusyTimeout):
+					logger.WithField("host", host).Warn("Upstream handlers busy, drop message")
```

Comment:
Shouldn't we register a drop event here?

Reply:
I'm not sure – at this point, there are no correlation IDs are this point, there's no corresponding `gs.up.receive` yet, so that's gonna be confusing.

We could consider a new event; `gs.up.fail`?